File: modules/dashboard/brightnessSlider.ts
import { Widget, Utils, Gio } from "imports";
import Brightness from "../service/brightness";

const { monitorFile, readFile, exec } = Utils
const { Box, Slider, Label } = Widget;

const BSlider = () =>
	Slider({
		className: "brightsld Slider",
		drawValue: false,
		on_change: self => Brightness.screen_value = self.value,
		value: Brightness.bind('screen-value').as(n => n > 1 ? 1 : n),
	});
	
const Icon = () =>
	Label({
		className: "brightsldIcon",
		setup: self => self.hook(Brightness, (self, screenValue) => {
        const icons = ["󰃚", "󰃛", "󰃜", "󰃝", "󰃞", "󰃟", "󰃠"];
        self.label =`${icons[Math.floor((Brightness.screen_value * 100) / 15)]}`;
    }, 'screen-changed'),
	});

export const BrightnessSlider = () =>
	Box({
		className: "brightSlider",
		vertical: true,
		hpack: "center",
		vpack: "center",
		children: [
			Label({
				className: "brightsldLabel",
				label: "Brightness",
				hpack: "center",
			}),
			Box({
				hpack: "center",
				vpack: "center",
				children: [
					Icon(),
					BSlider(),
				]
			}),

		],
	});

File: Rust/Modules/Widgets/BrightnessSlider.rs
use imports::Widget;
use crate::service::brightness;

fn b_slider() -> Widget {
    Slider {
        class_name: "brightsld Slider",
        draw_value: false,
        on_change: |self| brightness::screen_value = self.value,
        value: brightness::bind("screen-value").map(|n| if n > 1 { 1 } else { n }),
    }
}

fn icon() -> Widget {
    Label {
        class_name: "brightsldIcon",
        setup: |self| {
            self.hook(brightness, |self, screen_value| {
                let icons = ["󰃚", "󰃛", "󰃜", "󰃝", "󰃞", "󰃟", "󰃠"];
                self.label = format!("{}", icons[((brightness::screen_value * 100) / 15) as usize]);
            }, "screen-changed");
        },
    }
}

pub fn brightness_slider() -> Widget {
    Box {
        class_name: "brightSlider",
        vertical: true,
        hpack: "center",
        vpack: "center",
        children: vec![
            Label {
                class_name: "brightsldLabel",
                label: "Brightness",
                hpack: "center",
            },
            Box {
                hpack: "center",
                vpack: "center",
                children: vec![
                    icon(),
                    b_slider(),
                ],
            },
        ],
    }
}